File: packages/client-runtime/crates/bittery-client-bindings/facades/swift/BitteryClientRuntime.swift
import Foundation

/// Stable Swift facade; the raw UniFFI `shutdown` name avoids Kotlin's `AutoCloseable.close` clash.
public final class BitteryClientRuntime: @unchecked Sendable {
    private let native: ClientRuntime
    private let closeLock = NSLock()
    private var closed = false

    public init(native: ClientRuntime = ClientRuntime()) {
        self.native = native
    }

    public func request(_ request: RuntimeRequest) async throws -> RuntimeResponse {
        try await native.request(request: request)
    }

    public func observe(
        _ request: ObservationRequest,
        sink: ObservationSink
    ) throws -> ObservationHandle {
        try native.observe(request: request, sink: sink)
    }

    public func close() async {
        if beginClose() {
            await native.shutdown()
        }
    }

    private func beginClose() -> Bool {
        closeLock.lock()
        let shouldClose = !closed
        closed = true
        closeLock.unlock()
        return shouldClose
    }
}

File: packages/client-runtime/crates/bittery-client-bindings/facades/kotlin/BitteryClientRuntime.kt
package com.bittery.runtime

import java.util.concurrent.atomic.AtomicBoolean
import uniffi.bittery_client_bindings.ClientRuntime as NativeClientRuntime
import uniffi.bittery_client_bindings.ObservationHandle
import uniffi.bittery_client_bindings.ObservationRequest
import uniffi.bittery_client_bindings.ObservationSink
import uniffi.bittery_client_bindings.RuntimeRequest
import uniffi.bittery_client_bindings.RuntimeResponse

/** Stable Kotlin facade; the raw UniFFI `shutdown` name avoids its `AutoCloseable.close` collision. */
class BitteryClientRuntime(
    private val native: NativeClientRuntime = NativeClientRuntime(),
) {
    private val closed = AtomicBoolean(false)

    suspend fun request(request: RuntimeRequest): RuntimeResponse = native.request(request)

    fun observe(request: ObservationRequest, sink: ObservationSink): ObservationHandle =
        native.observe(request, sink)

    suspend fun close() {
        if (closed.compareAndSet(false, true)) {
            native.shutdown()
        }
    }
}
